In [1]:
import pandas as pd
import numpy as np

In [2]:
#Reading January 2019 dataset
jan = pd.read_csv('FLIGHT_ONTIME_REPORTING_JAN.csv')
jan.head()

QUARTER  MONTH    FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM ORIGIN_CITY_NAME  \
0        1      1  1/11/2019                  19393   N220WN      Orlando, FL   
1        1      1  1/11/2019                  19393   N7889A      Orlando, FL   
2        1      1  1/11/2019                  19393   N764SW      Orlando, FL   
3        1      1  1/11/2019                  19393   N8309C      Orlando, FL   
4        1      1  1/11/2019                  19393   N8686A      Orlando, FL   

  ORIGIN_STATE_NM  DEST_CITY_NAME DEST_STATE_NM  TAXI_OUT  ...  DIVERTED  \
0         Florida  Pittsburgh, PA  Pennsylvania       9.0  ...         0   
1         Florida  Pittsburgh, PA  Pennsylvania      14.0  ...         0   
2         Florida  Providence, RI  Rhode Island       9.0  ...         0   
3         Florida  Providence, RI  Rhode Island      15.0  ...         0   
4         Florida  Providence, RI  Rhode Island      10.0  ...         0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  CARRIER_DELAY  \
0                133.0     118.0        1       834            NaN   
1                138.0     119.0        1       834            NaN   
2                153.0     141.0        1      1072            NaN   
3                161.0     144.0        1      1072            NaN   
4                155.0     142.0        1      1072            NaN   

   WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0            NaN        NaN             NaN                  NaN  
1            NaN        NaN             NaN                  NaN  
2            NaN        NaN             NaN                  NaN  
3            NaN        NaN             NaN                  NaN  
4            NaN        NaN             NaN                  NaN  

[5 rows x 21 columns]

In [3]:
jan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 583985 entries, 0 to 583984
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   QUARTER                583985 non-null  int64  
 1   MONTH                  583985 non-null  int64  
 2   FL_DATE                583985 non-null  object 
 3   OP_CARRIER_AIRLINE_ID  583985 non-null  int64  
 4   TAIL_NUM               581442 non-null  object 
 5   ORIGIN_CITY_NAME       583985 non-null  object 
 6   ORIGIN_STATE_NM        583985 non-null  object 
 7   DEST_CITY_NAME         583985 non-null  object 
 8   DEST_STATE_NM          583985 non-null  object 
 9   TAXI_OUT               567369 non-null  float64
 10  TAXI_IN                566924 non-null  float64
 11  DIVERTED               583985 non-null  int64  
 12  ACTUAL_ELAPSED_TIME    565963 non-null  float64
 13  AIR_TIME               565963 non-null  float64
 14  FLIGHTS                583985 non-nu

- January,2019 dataset has 21 columns and 583985 rows.
- Some of the columns has missing values.

In [4]:
#Reading February 2019 dataset
feb = pd.read_csv("FLIGHT_ONTIME_REPORTING_FEB.csv")
feb.head()

QUARTER  MONTH   FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM ORIGIN_CITY_NAME  \
0        1      2  2/4/2019                  19805   N989AN        Miami, FL   
1        1      2  2/5/2019                  19805   N346PR        Miami, FL   
2        1      2  2/6/2019                  19805   N872NN        Miami, FL   
3        1      2  2/7/2019                  19805   N316PF        Miami, FL   
4        1      2  2/8/2019                  19805   N855NN        Miami, FL   

  ORIGIN_STATE_NM DEST_CITY_NAME DEST_STATE_NM  TAXI_OUT  ...  DIVERTED  \
0         Florida    Phoenix, AZ       Arizona      21.0  ...         0   
1         Florida    Phoenix, AZ       Arizona      22.0  ...         0   
2         Florida    Phoenix, AZ       Arizona      17.0  ...         0   
3         Florida    Phoenix, AZ       Arizona      13.0  ...         0   
4         Florida    Phoenix, AZ       Arizona      21.0  ...         0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  CARRIER_DELAY  \
0                300.0     274.0        1      1972            NaN   
1                330.0     298.0        1      1972            3.0   
2                316.0     290.0        1      1972            NaN   
3                319.0     298.0        1      1972            NaN   
4                310.0     283.0        1      1972            NaN   

   WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0            NaN        NaN             NaN                  NaN  
1            0.0       16.0             0.0                  0.0  
2            NaN        NaN             NaN                  NaN  
3            NaN        NaN             NaN                  NaN  
4            NaN        NaN             NaN                  NaN  

[5 rows x 21 columns]

In [5]:
feb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 533175 entries, 0 to 533174
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   QUARTER                533175 non-null  int64  
 1   MONTH                  533175 non-null  int64  
 2   FL_DATE                533175 non-null  object 
 3   OP_CARRIER_AIRLINE_ID  533175 non-null  int64  
 4   TAIL_NUM               531782 non-null  object 
 5   ORIGIN_CITY_NAME       533175 non-null  object 
 6   ORIGIN_STATE_NM        533175 non-null  object 
 7   DEST_CITY_NAME         533175 non-null  object 
 8   DEST_STATE_NM          533175 non-null  object 
 9   TAXI_OUT               518074 non-null  float64
 10  TAXI_IN                517475 non-null  float64
 11  DIVERTED               533175 non-null  int64  
 12  ACTUAL_ELAPSED_TIME    516314 non-null  float64
 13  AIR_TIME               516314 non-null  float64
 14  FLIGHTS                533175 non-nu

- February 2019 has 21 columns and 533175 rows with some missing values.

In [6]:
#Reading march 2019 dataset
mar = pd.read_csv("FLIGHT_ONTIME_REPORTING_MAR.csv")
mar.head()

QUARTER  MONTH    FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM ORIGIN_CITY_NAME  \
0        1      3  3/21/2019                  20363   N8943A    Knoxville, TN   
1        1      3  3/22/2019                  20363   N8688C    Knoxville, TN   
2        1      3  3/23/2019                  20363   N840AY    Knoxville, TN   
3        1      3  3/17/2019                  20363   N279PQ     New York, NY   
4        1      3  3/18/2019                  20363   N906XJ     New York, NY   

  ORIGIN_STATE_NM DEST_CITY_NAME DEST_STATE_NM  TAXI_OUT  ...  DIVERTED  \
0       Tennessee    Detroit, MI      Michigan      18.0  ...         0   
1       Tennessee    Detroit, MI      Michigan       9.0  ...         0   
2       Tennessee    Detroit, MI      Michigan      12.0  ...         0   
3        New York    Buffalo, NY      New York      22.0  ...         0   
4        New York    Buffalo, NY      New York      15.0  ...         0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  CARRIER_DELAY  \
0                100.0      73.0        1       443            NaN   
1                 99.0      76.0        1       443            NaN   
2                 91.0      71.0        1       443            NaN   
3                 81.0      56.0        1       301            NaN   
4                 84.0      59.0        1       301            NaN   

   WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0            NaN        NaN             NaN                  NaN  
1            NaN        NaN             NaN                  NaN  
2            NaN        NaN             NaN                  NaN  
3            NaN        NaN             NaN                  NaN  
4            NaN        NaN             NaN                  NaN  

[5 rows x 21 columns]

In [7]:
mar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 632074 entries, 0 to 632073
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   QUARTER                632074 non-null  int64  
 1   MONTH                  632074 non-null  int64  
 2   FL_DATE                632074 non-null  object 
 3   OP_CARRIER_AIRLINE_ID  632074 non-null  int64  
 4   TAIL_NUM               628006 non-null  object 
 5   ORIGIN_CITY_NAME       632074 non-null  object 
 6   ORIGIN_STATE_NM        632074 non-null  object 
 7   DEST_CITY_NAME         632074 non-null  object 
 8   DEST_STATE_NM          632074 non-null  object 
 9   TAXI_OUT               619606 non-null  float64
 10  TAXI_IN                619289 non-null  float64
 11  DIVERTED               632074 non-null  int64  
 12  ACTUAL_ELAPSED_TIME    618505 non-null  float64
 13  AIR_TIME               618505 non-null  float64
 14  FLIGHTS                632074 non-nu

- March 2019 dataset 63204 rows with some missing values.

In [8]:
#Reading april 2019 dataset
apr = pd.read_csv("FLIGHT_ONTIME_REPORTING_APR.csv")
apr.head()

QUARTER  MONTH    FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM ORIGIN_CITY_NAME  \
0        2      4  4/19/2019                  20416   N637NK  Minneapolis, MN   
1        2      4  4/20/2019                  20416   N907NK  Minneapolis, MN   
2        2      4  4/21/2019                  20416   N604NK  Minneapolis, MN   
3        2      4  4/22/2019                  20416   N642NK  Minneapolis, MN   
4        2      4  4/23/2019                  20416   N912NK  Minneapolis, MN   

  ORIGIN_STATE_NM DEST_CITY_NAME DEST_STATE_NM  TAXI_OUT  ...  DIVERTED  \
0       Minnesota      Tampa, FL       Florida      13.0  ...         0   
1       Minnesota      Tampa, FL       Florida      12.0  ...         0   
2       Minnesota      Tampa, FL       Florida       NaN  ...         0   
3       Minnesota      Tampa, FL       Florida      14.0  ...         0   
4       Minnesota      Tampa, FL       Florida      13.0  ...         0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  CARRIER_DELAY  \
0                195.0     171.0        1      1306            0.0   
1                168.0     148.0        1      1306            NaN   
2                  NaN       NaN        1      1306            NaN   
3                192.0     170.0        1      1306            NaN   
4                176.0     158.0        1      1306            NaN   

   WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0            0.0        3.0             0.0                317.0  
1            NaN        NaN             NaN                  NaN  
2            NaN        NaN             NaN                  NaN  
3            NaN        NaN             NaN                  NaN  
4            NaN        NaN             NaN                  NaN  

[5 rows x 21 columns]

In [9]:
apr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 612023 entries, 0 to 612022
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   QUARTER                612023 non-null  int64  
 1   MONTH                  612023 non-null  int64  
 2   FL_DATE                612023 non-null  object 
 3   OP_CARRIER_AIRLINE_ID  612023 non-null  int64  
 4   TAIL_NUM               609858 non-null  object 
 5   ORIGIN_CITY_NAME       612023 non-null  object 
 6   ORIGIN_STATE_NM        612023 non-null  object 
 7   DEST_CITY_NAME         612023 non-null  object 
 8   DEST_STATE_NM          612023 non-null  object 
 9   TAXI_OUT               597583 non-null  float64
 10  TAXI_IN                597341 non-null  float64
 11  DIVERTED               612023 non-null  int64  
 12  ACTUAL_ELAPSED_TIME    596020 non-null  float64
 13  AIR_TIME               596020 non-null  float64
 14  FLIGHTS                612023 non-nu

- April dataset has a total of 612023 rows with some mising values.

In [10]:
#Reading may 2019 dataset
may = pd.read_csv('FLIGHT_ONTIME_REPORTING_MAY.csv')
may.head()

QUARTER  MONTH    FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM ORIGIN_CITY_NAME  \
0        2      5  5/28/2019                  20452   N732YX    St. Louis, MO   
1        2      5  5/28/2019                  20452   N652RW      Chicago, IL   
2        2      5  5/28/2019                  20452   N862RW  Minneapolis, MN   
3        2      5  5/28/2019                  20452   N862RW      Chicago, IL   
4        2      5  5/28/2019                  20452   N732YX       Newark, NJ   

  ORIGIN_STATE_NM      DEST_CITY_NAME   DEST_STATE_NM  TAXI_OUT  ...  \
0        Missouri          Newark, NJ      New Jersey      23.0  ...   
1        Illinois  Raleigh/Durham, NC  North Carolina      12.0  ...   
2       Minnesota         Chicago, IL        Illinois      16.0  ...   
3        Illinois       Nashville, TN       Tennessee      31.0  ...   
4      New Jersey     Kansas City, MO        Missouri      17.0  ...   

   DIVERTED  ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  CARRIER_DELAY  \
0         0                151.0     117.0        1       872            NaN   
1         0                109.0      89.0        1       646            NaN   
2         0                100.0      54.0        1       334            NaN   
3         0                 98.0      61.0        1       409            NaN   
4         0                180.0     157.0        1      1092            NaN   

   WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0            NaN        NaN             NaN                  NaN  
1            NaN        NaN             NaN                  NaN  
2            NaN        NaN             NaN                  NaN  
3            NaN        NaN             NaN                  NaN  
4            NaN        NaN             NaN                  NaN  

[5 rows x 21 columns]

In [11]:
may.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636390 entries, 0 to 636389
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   QUARTER                636390 non-null  int64  
 1   MONTH                  636390 non-null  int64  
 2   FL_DATE                636390 non-null  object 
 3   OP_CARRIER_AIRLINE_ID  636390 non-null  int64  
 4   TAIL_NUM               635272 non-null  object 
 5   ORIGIN_CITY_NAME       636390 non-null  object 
 6   ORIGIN_STATE_NM        636390 non-null  object 
 7   DEST_CITY_NAME         636390 non-null  object 
 8   DEST_STATE_NM          636390 non-null  object 
 9   TAXI_OUT               623438 non-null  float64
 10  TAXI_IN                623166 non-null  float64
 11  DIVERTED               636390 non-null  int64  
 12  ACTUAL_ELAPSED_TIME    621339 non-null  float64
 13  AIR_TIME               621339 non-null  float64
 14  FLIGHTS                636390 non-nu

- May dataset has a total of 636390 rows and some of the columns has missing values.

In [12]:
#Reading june 2019 dataset
jun = pd.read_csv("FLIGHT_ONTIME_REPORTING_JUN.csv")
jun.head()

QUARTER  MONTH   FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM ORIGIN_CITY_NAME  \
0        2      6  6/1/2019                  20409   N929JB    San Diego, CA   
1        2      6  6/1/2019                  20409   N981JT     New York, NY   
2        2      6  6/1/2019                  20409   N969JT  Los Angeles, CA   
3        2      6  6/1/2019                  20409   N339JB    Baltimore, MD   
4        2      6  6/1/2019                  20409   N997JL      Orlando, FL   

  ORIGIN_STATE_NM   DEST_CITY_NAME  DEST_STATE_NM  TAXI_OUT  ...  DIVERTED  \
0      California       Boston, MA  Massachusetts      18.0  ...         0   
1        New York  Los Angeles, CA     California      14.0  ...         0   
2      California     New York, NY       New York      12.0  ...         0   
3        Maryland       Boston, MA  Massachusetts      16.0  ...         0   
4         Florida    Aguadilla, PR    Puerto Rico      13.0  ...         0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  CARRIER_DELAY  \
0                319.0     296.0        1      2588            NaN   
1                363.0     343.0        1      2475            NaN   
2                322.0     300.0        1      2475            NaN   
3                 79.0      59.0        1       369            NaN   
4                144.0     127.0        1      1129           14.0   

   WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0            NaN        NaN             NaN                  NaN  
1            NaN        NaN             NaN                  NaN  
2            NaN        NaN             NaN                  NaN  
3            NaN        NaN             NaN                  NaN  
4            0.0        0.0             0.0                 39.0  

[5 rows x 21 columns]

In [13]:
jun.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636691 entries, 0 to 636690
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   QUARTER                636691 non-null  int64  
 1   MONTH                  636691 non-null  int64  
 2   FL_DATE                636691 non-null  object 
 3   OP_CARRIER_AIRLINE_ID  636691 non-null  int64  
 4   TAIL_NUM               635337 non-null  object 
 5   ORIGIN_CITY_NAME       636691 non-null  object 
 6   ORIGIN_STATE_NM        636691 non-null  object 
 7   DEST_CITY_NAME         636691 non-null  object 
 8   DEST_STATE_NM          636691 non-null  object 
 9   TAXI_OUT               623542 non-null  float64
 10  TAXI_IN                623233 non-null  float64
 11  DIVERTED               636691 non-null  int64  
 12  ACTUAL_ELAPSED_TIME    620887 non-null  float64
 13  AIR_TIME               620887 non-null  float64
 14  FLIGHTS                636691 non-nu

- June has a total of 636691 rows with some missing values.

In [14]:
#Joining the jan to june 2019 datas into a single dataset
months = [jan,feb,mar,apr,may,jun]
df = pd.concat(months,ignore_index = True)
df.head()

QUARTER  MONTH    FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM ORIGIN_CITY_NAME  \
0        1      1  1/11/2019                  19393   N220WN      Orlando, FL   
1        1      1  1/11/2019                  19393   N7889A      Orlando, FL   
2        1      1  1/11/2019                  19393   N764SW      Orlando, FL   
3        1      1  1/11/2019                  19393   N8309C      Orlando, FL   
4        1      1  1/11/2019                  19393   N8686A      Orlando, FL   

  ORIGIN_STATE_NM  DEST_CITY_NAME DEST_STATE_NM  TAXI_OUT  ...  DIVERTED  \
0         Florida  Pittsburgh, PA  Pennsylvania       9.0  ...         0   
1         Florida  Pittsburgh, PA  Pennsylvania      14.0  ...         0   
2         Florida  Providence, RI  Rhode Island       9.0  ...         0   
3         Florida  Providence, RI  Rhode Island      15.0  ...         0   
4         Florida  Providence, RI  Rhode Island      10.0  ...         0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  CARRIER_DELAY  \
0                133.0     118.0        1       834            NaN   
1                138.0     119.0        1       834            NaN   
2                153.0     141.0        1      1072            NaN   
3                161.0     144.0        1      1072            NaN   
4                155.0     142.0        1      1072            NaN   

   WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0            NaN        NaN             NaN                  NaN  
1            NaN        NaN             NaN                  NaN  
2            NaN        NaN             NaN                  NaN  
3            NaN        NaN             NaN                  NaN  
4            NaN        NaN             NaN                  NaN  

[5 rows x 21 columns]

In [15]:
df.shape

(3634338, 21)

- The combination of the first half of the 2019 dataset has a total of 3634338 rows and 21 columns

In [16]:
#Checking the basic statistical information of the dataset
df.describe()

QUARTER         MONTH  OP_CARRIER_AIRLINE_ID      TAXI_OUT  \
count  3.634338e+06  3.634338e+06           3.634338e+06  3.549612e+06   
mean   1.518693e+00  3.576097e+00           1.998447e+04  1.751871e+01   
std    4.996505e-01  1.703839e+00           3.756723e+02  1.015717e+01   
min    1.000000e+00  1.000000e+00           1.939300e+04  1.000000e+00   
25%    1.000000e+00  2.000000e+00           1.979000e+04  1.100000e+01   
50%    2.000000e+00  4.000000e+00           1.997700e+04  1.500000e+01   
75%    2.000000e+00  5.000000e+00           2.036800e+04  2.000000e+01   
max    2.000000e+00  6.000000e+00           2.045200e+04  2.270000e+02   

            TAXI_IN      DIVERTED  ACTUAL_ELAPSED_TIME      AIR_TIME  \
count  3.547428e+06  3.634338e+06         3.539028e+06  3.539028e+06   
mean   7.660759e+00  2.761989e-03         1.376193e+02  1.124545e+02   
std    6.122517e+00  5.248200e-02         7.315557e+01  7.112862e+01   
min    1.000000e+00  0.000000e+00         1.500000e+01  4.000000e+00   
25%    4.000000e+00  0.000000e+00         8.400000e+01  6.100000e+01   
50%    6.000000e+00  0.000000e+00         1.200000e+02  9.400000e+01   
75%    9.000000e+00  0.000000e+00         1.680000e+02  1.420000e+02   
max    3.160000e+02  1.000000e+00         1.604000e+03  1.557000e+03   

         FLIGHTS      DISTANCE  CARRIER_DELAY  WEATHER_DELAY      NAS_DELAY  \
count  3634338.0  3.634338e+06  719135.000000  719135.000000  719135.000000   
mean         1.0  8.012517e+02      20.922062       4.106054      16.633123   
std          0.0  5.914274e+02      64.556750      33.886872      39.018400   
min          1.0  3.100000e+01       0.000000       0.000000       0.000000   
25%          1.0  3.690000e+02       0.000000       0.000000       0.000000   
50%          1.0  6.410000e+02       0.000000       0.000000       3.000000   
75%          1.0  1.036000e+03      18.000000       0.000000      20.000000   
max          1.0  5.095000e+03    2592.000000    1521.000000    1741.000000   

       SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
count   719135.000000        719135.000000  
mean         0.108811            27.716438  
std          4.093113            53.689881  
min          0.000000             0.000000  
25%          0.000000             0.000000  
50%          0.000000             3.000000  
75%          0.000000            34.000000  
max       1078.000000          2206.000000

In [17]:
df.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3634338 entries, 0 to 3634337
Data columns (total 21 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   QUARTER                3634338 non-null  int64  
 1   MONTH                  3634338 non-null  int64  
 2   FL_DATE                3634338 non-null  object 
 3   OP_CARRIER_AIRLINE_ID  3634338 non-null  int64  
 4   TAIL_NUM               3621697 non-null  object 
 5   ORIGIN_CITY_NAME       3634338 non-null  object 
 6   ORIGIN_STATE_NM        3634338 non-null  object 
 7   DEST_CITY_NAME         3634338 non-null  object 
 8   DEST_STATE_NM          3634338 non-null  object 
 9   TAXI_OUT               3549612 non-null  float64
 10  TAXI_IN                3547428 non-null  float64
 11  DIVERTED               3634338 non-null  int64  
 12  ACTUAL_ELAPSED_TIME    3539028 non-null  float64
 13  AIR_TIME               3539028 non-null  float64
 14  FLIGHTS           

In [18]:
(df.isnull().sum()/len(df)).round(2) * 100

QUARTER                   0.0
MONTH                     0.0
FL_DATE                   0.0
OP_CARRIER_AIRLINE_ID     0.0
TAIL_NUM                  0.0
ORIGIN_CITY_NAME          0.0
ORIGIN_STATE_NM           0.0
DEST_CITY_NAME            0.0
DEST_STATE_NM             0.0
TAXI_OUT                  2.0
TAXI_IN                   2.0
DIVERTED                  0.0
ACTUAL_ELAPSED_TIME       3.0
AIR_TIME                  3.0
FLIGHTS                   0.0
DISTANCE                  0.0
CARRIER_DELAY            80.0
WEATHER_DELAY            80.0
NAS_DELAY                80.0
SECURITY_DELAY           80.0
LATE_AIRCRAFT_DELAY      80.0
dtype: float64

- The combined data has about 2% missing data in the Taxi out column and Taxi in columns respectively,3% in the Air time column and about 80% missing in each of the Delay columns.

###### AIRLINE WITH MOST NUMBER OF FLIGHT IN FIRST QUATER OF 2019

In [19]:
#Datafram for first quarter of 2019
q1 = df[df['QUARTER'] == 1]
q1.head()

QUARTER  MONTH    FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM ORIGIN_CITY_NAME  \
0        1      1  1/11/2019                  19393   N220WN      Orlando, FL   
1        1      1  1/11/2019                  19393   N7889A      Orlando, FL   
2        1      1  1/11/2019                  19393   N764SW      Orlando, FL   
3        1      1  1/11/2019                  19393   N8309C      Orlando, FL   
4        1      1  1/11/2019                  19393   N8686A      Orlando, FL   

  ORIGIN_STATE_NM  DEST_CITY_NAME DEST_STATE_NM  TAXI_OUT  ...  DIVERTED  \
0         Florida  Pittsburgh, PA  Pennsylvania       9.0  ...         0   
1         Florida  Pittsburgh, PA  Pennsylvania      14.0  ...         0   
2         Florida  Providence, RI  Rhode Island       9.0  ...         0   
3         Florida  Providence, RI  Rhode Island      15.0  ...         0   
4         Florida  Providence, RI  Rhode Island      10.0  ...         0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  CARRIER_DELAY  \
0                133.0     118.0        1       834            NaN   
1                138.0     119.0        1       834            NaN   
2                153.0     141.0        1      1072            NaN   
3                161.0     144.0        1      1072            NaN   
4                155.0     142.0        1      1072            NaN   

   WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0            NaN        NaN             NaN                  NaN  
1            NaN        NaN             NaN                  NaN  
2            NaN        NaN             NaN                  NaN  
3            NaN        NaN             NaN                  NaN  
4            NaN        NaN             NaN                  NaN  

[5 rows x 21 columns]

In [20]:
q1.groupby('OP_CARRIER_AIRLINE_ID')['FLIGHTS'].sum().sort_values(ascending=False)

OP_CARRIER_AIRLINE_ID
19393    330225
19805    228103
19790    225381
20304    194934
19977    144288
20452     76818
20398     75751
20409     72768
20397     69080
19930     61466
20363     60880
20378     53701
20416     46030
20366     37584
20436     28242
20368     24294
19690     19689
Name: FLIGHTS, dtype: int64

- Airline with ID 19393 has the highest number of flights(330225 flights) in first quarter of 2019

###### TOTAL NUMBER OF AIR TRAVELS IN THE FIRST HALF OF 2019

In [21]:
df['FLIGHTS'].sum()

3634338

- The total number of airline travels in the first half of 2019 is 3634338

###### TOTAL AIRTIME BY ALL AIRLINES IN MAY 2019

In [22]:
may['AIR_TIME'].sum()/60

1154951.7333333334

- The total airtime by airlines in may 2019 is 1154951.7 hours

###### THE BUSIEST CITY IN APRIL 2019

In [23]:
df.columns

Index(['QUARTER', 'MONTH', 'FL_DATE', 'OP_CARRIER_AIRLINE_ID', 'TAIL_NUM',
       'ORIGIN_CITY_NAME', 'ORIGIN_STATE_NM', 'DEST_CITY_NAME',
       'DEST_STATE_NM', 'TAXI_OUT', 'TAXI_IN', 'DIVERTED',
       'ACTUAL_ELAPSED_TIME', 'AIR_TIME', 'FLIGHTS', 'DISTANCE',
       'CARRIER_DELAY', 'WEATHER_DELAY', 'NAS_DELAY', 'SECURITY_DELAY',
       'LATE_AIRCRAFT_DELAY'],
      dtype='object')

In [24]:
apr.groupby('ORIGIN_CITY_NAME')['FLIGHTS'].count().sort_values(ascending=False).head(1)

ORIGIN_CITY_NAME
Chicago, IL    34396
Name: FLIGHTS, dtype: int64

In [25]:
apr['DEST_CITY_NAME'].value_counts().head()

Chicago, IL              34393
Atlanta, GA              32941
New York, NY             25095
Dallas/Fort Worth, TX    23930
Houston, TX              19941
Name: DEST_CITY_NAME, dtype: int64

- Chicago, IL appears to be the busiest city in April 2019

###### AIRLINE WITH THE MOST DISTANCE COVERED IN JUNE 2019

In [26]:
jun.groupby('OP_CARRIER_AIRLINE_ID')['DISTANCE'].sum().sort_values(ascending=False)

OP_CARRIER_AIRLINE_ID
19393    85246573
19805    80947454
19790    79641435
19977    65963246
20304    35631686
19930    30800766
20409    26972049
20416    17576612
20452    16804684
20398    12910059
20436    11879495
20378    11066389
20363    10009285
20397     9877331
20368     9283200
19690     5316126
20366     4389648
Name: DISTANCE, dtype: int64

- SOUTHWEST AIRLINES CO. WN seems to covered the most distance in june 2019.

##### THE TAIL NUMBER OF THE AIRCRAFT THAT COVERED THE MOST DISTANCE IN SECOND QUATER OF 2019

In [27]:
q2 = df[df['QUARTER'] == 2]
q2.head()

QUARTER  MONTH    FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM  \
1749234        2      4  4/19/2019                  20416   N637NK   
1749235        2      4  4/20/2019                  20416   N907NK   
1749236        2      4  4/21/2019                  20416   N604NK   
1749237        2      4  4/22/2019                  20416   N642NK   
1749238        2      4  4/23/2019                  20416   N912NK   

        ORIGIN_CITY_NAME ORIGIN_STATE_NM DEST_CITY_NAME DEST_STATE_NM  \
1749234  Minneapolis, MN       Minnesota      Tampa, FL       Florida   
1749235  Minneapolis, MN       Minnesota      Tampa, FL       Florida   
1749236  Minneapolis, MN       Minnesota      Tampa, FL       Florida   
1749237  Minneapolis, MN       Minnesota      Tampa, FL       Florida   
1749238  Minneapolis, MN       Minnesota      Tampa, FL       Florida   

         TAXI_OUT  ...  DIVERTED  ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  \
1749234      13.0  ...         0                195.0     171.0        1   
1749235      12.0  ...         0                168.0     148.0        1   
1749236       NaN  ...         0                  NaN       NaN        1   
1749237      14.0  ...         0                192.0     170.0        1   
1749238      13.0  ...         0                176.0     158.0        1   

         DISTANCE  CARRIER_DELAY  WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  \
1749234      1306            0.0            0.0        3.0             0.0   
1749235      1306            NaN            NaN        NaN             NaN   
1749236      1306            NaN            NaN        NaN             NaN   
1749237      1306            NaN            NaN        NaN             NaN   
1749238      1306            NaN            NaN        NaN             NaN   

         LATE_AIRCRAFT_DELAY  
1749234                317.0  
1749235                  NaN  
1749236                  NaN  
1749237                  NaN  
1749238                  NaN  

[5 rows x 21 columns]

In [28]:
q2.groupby(['TAIL_NUM','OP_CARRIER_AIRLINE_ID'])['DISTANCE'].sum().sort_values(ascending=False).head()

TAIL_NUM  OP_CARRIER_AIRLINE_ID
N934JB    20409                    598286
N943JT    20409                    592006
N984JB    20409                    588931
N982JB    20409                    585632
N979JT    20409                    583407
Name: DISTANCE, dtype: int64

- Aircraft with tail number N934JB covered the most distance in second quarter of 2019

###### STATE WITH THE MOST DELAYS IN SECOND QUARTER OF 2019

In [29]:
q2['DELAY'] = q2['CARRIER_DELAY']+q2['WEATHER_DELAY']+q2['NAS_DELAY']+ q2['SECURITY_DELAY']+q2['LATE_AIRCRAFT_DELAY']

<ipython-input-29-cc3857476fa4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  q2['DELAY'] = q2['CARRIER_DELAY']+q2['WEATHER_DELAY']+q2['NAS_DELAY']+ q2['SECURITY_DELAY']+q2['LATE_AIRCRAFT_DELAY']


In [30]:
q2.tail()

QUARTER  MONTH    FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM  \
3634333        2      6  6/18/2019                  19805   N737US   
3634334        2      6  6/19/2019                  19805   N8001N   
3634335        2      6  6/20/2019                  19805   N9023N   
3634336        2      6  6/21/2019                  19805   N974AN   
3634337        2      6  6/22/2019                  19805   N4032T   

        ORIGIN_CITY_NAME ORIGIN_STATE_NM DEST_CITY_NAME DEST_STATE_NM  \
3634333      Chicago, IL        Illinois   Columbus, OH          Ohio   
3634334      Chicago, IL        Illinois   Columbus, OH          Ohio   
3634335      Chicago, IL        Illinois   Columbus, OH          Ohio   
3634336      Chicago, IL        Illinois   Columbus, OH          Ohio   
3634337      Chicago, IL        Illinois   Columbus, OH          Ohio   

         TAXI_OUT  ...  ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  \
3634333      38.0  ...                 93.0      48.0        1       296   
3634334      24.0  ...                 79.0      50.0        1       296   
3634335      23.0  ...                 78.0      49.0        1       296   
3634336      14.0  ...                 67.0      46.0        1       296   
3634337      16.0  ...                 65.0      46.0        1       296   

         CARRIER_DELAY  WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  \
3634333            NaN            NaN        NaN             NaN   
3634334            NaN            NaN        NaN             NaN   
3634335            NaN            NaN        NaN             NaN   
3634336           44.0            0.0        0.0             0.0   
3634337            NaN            NaN        NaN             NaN   

         LATE_AIRCRAFT_DELAY  DELAY  
3634333                  NaN    NaN  
3634334                  NaN    NaN  
3634335                  NaN    NaN  
3634336                  0.0   44.0  
3634337                  NaN    NaN  

[5 rows x 22 columns]

In [31]:
q2.groupby('ORIGIN_STATE_NM')['DELAY'].sum().sort_values(ascending=False).head()

ORIGIN_STATE_NM
Texas         3619582.0
Florida       2510825.0
California    2232886.0
Illinois      2017533.0
New York      1815198.0
Name: DELAY, dtype: float64

- Texas recorded the most delays in second qaurter of 2019 

###### DAILY AVERAGE TIME OUT IN NEW YORK IN FEBRUARY 2019

In [32]:
feb_ny = feb[feb['ORIGIN_STATE_NM'] == 'New York']
feb_ny.head()

QUARTER  MONTH    FL_DATE  OP_CARRIER_AIRLINE_ID TAIL_NUM  \
490        1      2  2/14/2019                  19805   N352PS   
491        1      2  2/15/2019                  19805   N871NN   
492        1      2  2/16/2019                  19805   N867NN   
493        1      2  2/17/2019                  19805   N985NN   
494        1      2  2/18/2019                  19805   N855NN   

    ORIGIN_CITY_NAME ORIGIN_STATE_NM DEST_CITY_NAME DEST_STATE_NM  TAXI_OUT  \
490     New York, NY        New York  Las Vegas, NV        Nevada      16.0   
491     New York, NY        New York  Las Vegas, NV        Nevada      17.0   
492     New York, NY        New York  Las Vegas, NV        Nevada      15.0   
493     New York, NY        New York  Las Vegas, NV        Nevada      18.0   
494     New York, NY        New York  Las Vegas, NV        Nevada      56.0   

     ...  DIVERTED  ACTUAL_ELAPSED_TIME  AIR_TIME  FLIGHTS  DISTANCE  \
490  ...         0                386.0     343.0        1      2248   
491  ...         0                355.0     330.0        1      2248   
492  ...         0                337.0     315.0        1      2248   
493  ...         0                349.0     321.0        1      2248   
494  ...         0                425.0     351.0        1      2248   

     CARRIER_DELAY  WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  \
490            4.0            0.0       19.0             0.0   
491            7.0            0.0        0.0             0.0   
492            NaN            NaN        NaN             NaN   
493            NaN            NaN        NaN             NaN   
494            0.0            0.0       56.0             0.0   

     LATE_AIRCRAFT_DELAY  
490                 22.0  
491                 40.0  
492                  NaN  
493                  NaN  
494                  0.0  

[5 rows x 21 columns]

In [33]:
feb_ny['TAXI_OUT'].mean()

24.689726862140656

- The average daily taxi out time in New York in february 2019 is 24.689

##### AIRLINES WITH NEW YORK DESTINATION THAT WERE DIVERTED

In [34]:
q1_div = q1[q1['DIVERTED'] == 1]

In [35]:
q1_div.shape

(3907, 21)

In [36]:
q1_div[q1_div['DEST_STATE_NM'] == 'New York']['DIVERTED'].sum()

260

- 260  airlines with new york bound destination were diverted

######  MOST RAMPANT DELAY IN WASHINTON IN THE FIRST HALF OF 2019

In [37]:
df[df["ORIGIN_STATE_NM"] == 'Washington'][['CARRIER_DELAY','WEATHER_DELAY','NAS_DELAY','SECURITY_DELAY','LATE_AIRCRAFT_DELAY']].sum().sort_values(ascending=False)

LATE_AIRCRAFT_DELAY    242206.0
NAS_DELAY              238556.0
CARRIER_DELAY          230277.0
WEATHER_DELAY           59108.0
SECURITY_DELAY            988.0
dtype: float64

- Late aircraft delay is the most rampant delay in first half of 2019 